File: src/main/java/com/falsepattern/zigbrains/zon/formatter/ZonFormattingModelBuilder.java
/*
 * Copyright 2023 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.zon.formatter;

import com.falsepattern.zigbrains.zon.ZonLanguage;
import com.intellij.formatting.FormattingContext;
import com.intellij.formatting.FormattingModel;
import com.intellij.formatting.FormattingModelBuilder;
import com.intellij.formatting.FormattingModelProvider;
import com.intellij.formatting.SpacingBuilder;
import com.intellij.psi.codeStyle.CodeStyleSettings;
import org.jetbrains.annotations.NotNull;

public class ZonFormattingModelBuilder implements FormattingModelBuilder {
    private static SpacingBuilder createSpaceBuilder(CodeStyleSettings settings) {
        return new SpacingBuilder(settings, ZonLanguage.INSTANCE);
    }
    @Override
    public @NotNull FormattingModel createModel(@NotNull FormattingContext formattingContext) {
        final var codeStyleSettings = formattingContext.getCodeStyleSettings();
        return FormattingModelProvider
                .createFormattingModelForPsiFile(formattingContext.getContainingFile(),
                                                 new ZonBlock(formattingContext.getNode(),
                                                              null,
                                                              null,
                                                              createSpaceBuilder(codeStyleSettings)),
                                                 codeStyleSettings);
    }
}

File: modules/core/src/main/kotlin/com/falsepattern/zigbrains/zon/formatter/ZonFormattingModelBuilder.kt
package com.falsepattern.zigbrains.zon.formatter

import com.falsepattern.zigbrains.zon.ZonLanguage
import com.intellij.formatting.FormattingContext
import com.intellij.formatting.FormattingModel
import com.intellij.formatting.FormattingModelBuilder
import com.intellij.formatting.FormattingModelProvider
import com.intellij.formatting.SpacingBuilder

class ZonFormattingModelBuilder : FormattingModelBuilder {
    override fun createModel(context: FormattingContext): FormattingModel {
        val settings = context.codeStyleSettings
        return FormattingModelProvider.createFormattingModelForPsiFile(
            context.containingFile,
            ZonBlock(context.node, null, null, SpacingBuilder(settings, ZonLanguage)),
            settings
        )
    }
}
